src/fldmod_by_const.jl

```diff
@@ -0,0 +1,147 @@
+twoToTheSizeOf(::Type{BigInt}) = BigInt(2)^256
+
+_leading_zeros(x) = leading_zeros(x)
+_leading_zeros(x::BigInt) = leading_zeros((x >> 128) % UInt128)
+
+
+# Splits `x` into `(hi,lo)` and returns them as the corresponding narrowed type
+function splitint(x::T) where {T<:Integer}
+    NT = narrow(T)
+    (x >> nbits(NT)) % NT, x % NT
+end
+
+narrow(::Type{Int128}) = Int64
+narrow(::Type{Int64}) = Int32
+narrow(::Type{Int32}) = Int16
+narrow(::Type{Int16}) = Int8
+narrow(::Type{UInt128}) = UInt64
+narrow(::Type{UInt64}) = UInt32
+narrow(::Type{UInt32}) = UInt16
+narrow(::Type{UInt16}) = UInt8
+
+"""    nbits(x) = sizeof(x)*8 """
+nbits(x) = sizeof(x)*8
+
+# Implemenation based on umul32hi, from https://stackoverflow.com/a/22847373/751061
+# /* compute the upper 32 bits of the product of two unsigned 32-bit integers */
+@inline function unsigned_splitmul_upper(a::T, b::T) where T<:Unsigned
+    # /* split operands into halves */
+    ah,al = splitint(a)
+    bh,bl = splitint(b)
+    halfT = typeof(ah)
+    halfbits = nbits(al)
+    # /* compute partial products */
+    p0 = widemul(al, bl);
+    p1 = widemul(al, bh);
+    p2 = widemul(ah, bl);
+    p3 = widemul(ah, bh);
+    # /* sum partial products */
+    cy = ((p0 >> halfbits) + (p1%halfT) + (p2%halfT)) >> halfbits;
+    return p3 + (p2 >> halfbits) + (p1 >> halfbits) + cy;
+end
+@inline function splitmul_upper(a::T, b::T) where T<:Unsigned
+    return unsigned_splitmul_upper(a,b)
+end
+@inline function splitmul_upper(a::T, b::T) where T<:Signed
+    return signed(unsigned_splitmul_upper(unsigned(a), unsigned(b))) - ((a < 0) ? b : 0) - ((b < 0) ? a : 0);
+end
+@inline splitmul_upper(a,b) = splitmul_upper(promote(a,b)...)
+
+
+"""
+    div_by_const(x, Val(C))
+
+The idea behind `div_by_const(x, Val(C))` is that we can avoid division entirely if
+instead of dividing by `C`, we multiply by `(2^N/C)`, then "divide by" `2^N`.
+
+And if we choose `N` to be `nbits(T)`, we can avoid the "divide by `2^N`" by just taking the
+_upper half_ of the result of the multiplication. And so, to do the multiplication, we use
+a `splitwidemul` which avoids widening by doing splitting x into upper and lower bits, doing
+several smaller multiplies, and then returning the result as two Ts, for the up and lo.
+
+REQUIRES:
+  - `C` _must be_ greater than `0`
+"""
+function div_by_const(x::T, ::Val{C}) where {T, C}
+# These checks will be compiled away during specialization.
+    if C == 1
+        return x
+    elseif (C <= 0)
+        throw(DomainError("C must be > 0"))
+    end
+    inv_coeff, toshift = calculate_inv_coeff(T, C)
+    up = splitmul_upper(x, inv_coeff)
+    out = up    # By keeping only the upper half, we're essentially dividing by 2^nbits(T)
+    # This condition will be compiled away during specialization.
+    if T <: Signed
+        # Because our magic number has a leading one, the result is negative if it's Signed.
+        # We add x to give us the positive equivalent.
+        out += x
```

Comment:
Can you give an example for 'out += x' giving the positive equivalent?  I'm a bit sleepy and having trouble wrapping my mind around it.

Reply:
lol i dunno, i'm getting conflicting information.

I _think_ that `movq` is `rax = rdx`, not the other way around. (Also, i would hope llvm wouldn't be generating useless instructions?)

Haha this pdf:
https://www.cs.cmu.edu/~fp/courses/15213-s07/misc/asm64-handout.pdf
says it's movq Source, Destination:
```movq S, D | D ← S  | Move quad word```

But THIS site:
https://www.felixcloutier.com/x86/MOVD:MOVQ.html
says the opposite:
> Copies a doubleword from the source operand (second operand) to the destination operand (first operand).

So lol i dunno.

This doc agrees with the first one:
https://cs.brown.edu/courses/cs033/docs/guides/x64_cheatsheet.pdf

And so does this one:
https://wiki.cdot.senecacollege.ca/wiki/X86_64_Register_and_Instruction_Quick_Start

So i'm pretty sure it's `movq Source, Destination`.

----

> Does your code work correctly if you drop the 'out += x' line?

And no, the tests fail if I remove this line. I'd sure like to know what it does, though. :(

Reply:
> What does `@code_llvm` show on your machine?

The trouble is the optimization happens after the LLVM, during native assembly generation:
```julia
julia> @code_llvm f(Int64(3))
define i64 @julia_f_35551(i64) {
top:
  %1 = sdiv i64 %0, 100
  ret i64 %1
}

julia> @code_llvm f(UInt64(3))
define i64 @julia_f_35563(i64) {
top:
  %1 = udiv i64 %0, 100
  ret i64 %1
}
```